src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,275 @@
         }
 
+        private async Task<PingReply> SendIcmpEchoRequestOverRawSocket(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        {
+            EndPoint endPoint = new IPEndPoint(address, 0);
+
+            bool isIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            ProtocolType protocolType = isIpv4 ? ProtocolType.Icmp : ProtocolType.IcmpV6;
+            // Use the current thread's ID as the identifier.
+            ushort identifier = (ushort)Environment.CurrentManagedThreadId;
+            IcmpHeader header = new IcmpHeader()
+            {
+                Type = isIpv4 ? (byte)Icmpv4MessageConstants.EchoRequest : (byte)Icmpv6MessageConstants.EchoRequest,
+                Code = 0,
+                HeaderChecksum = 0,
+                Identifier = identifier,
+                SequenceNumber = 0,
+            };
+            byte[] sendBuffer = CreateSendMessageBuffer(header, buffer);
+
+            using (Socket socket = new Socket(address.AddressFamily, SocketType.Raw, protocolType))
+            {
+                socket.ReceiveTimeout = timeout;
+                socket.SendTimeout = timeout;
+                // Setting Socket.DontFragment and .Ttl is not supported on Unix, so ignore the PingOptions parameter.
+
+                int ipHeaderLength = isIpv4 ? IpHeaderLengthInBytes : 0;
+                await socket.SendToAsync(new ArraySegment<byte>(sendBuffer), SocketFlags.None, endPoint).ConfigureAwait(false);
+                byte[] receiveBuffer = new byte[ipHeaderLength + IcmpHeaderLengthInBytes + buffer.Length];
+
+                long elapsed;
+                Stopwatch sw = Stopwatch.StartNew();
+                // Read from the socket in a loop. We may receive messages that are not echo replies, or that are not in response
+                // to the echo request we just sent. We need to filter such messages out, and continue reading until our timeout.
+                // For example, when pinging the local host, we need to filter out our own echo requests that the socket reads.
+                while ((elapsed = sw.ElapsedMilliseconds) < timeout)
+                {
+                    Task<SocketReceiveFromResult> receiveTask = socket.ReceiveFromAsync(
+                                                                    new ArraySegment<byte>(receiveBuffer),
+                                                                    SocketFlags.None,
+                                                                    endPoint);
+                    Task finished = await Task.WhenAny(receiveTask, Task.Delay(timeout - (int)elapsed)).ConfigureAwait(false);
+                    if (finished != receiveTask)
+                    {
+                        sw.Stop();
+                        throw new PingException(SR.net_ping_reply_timeout);
+                    }
+
+                    SocketReceiveFromResult receiveResult = receiveTask.Result;
```

Comment:
I'd suggest using ```receiveTask.GetAwaiter().GetResult()``` instead, so that if there is an exception, it'll propagate out directly rather than wrapped in the AggregateException that Result will propagate.